### PR TITLE
Precompute constants for batch normalization 

### PR DESCRIPTION
**Pull request by lm-konda**
## Motivation and Context
This PR is based on #133 

We can pre-compute `scale` and `shift` which is used in batch normalization.
So implement and add new optimization pass. 

## Description
　@antonionevado and @nlpng and I implement this PR.
　calculate `shift` and `scale` from `gamma`, `beta`, `mean`, `variance`, `epsilon` in advance.
　and make template file to fit new interface.
　and separate .cpp and .h for batch norm for readability.

modified:
　dlk/python/dlk/core/operators.py
　dlk/python/dlk/core/optimizer.py
　dlk/python/dlk/core/view.py
　dlk/python/dlk/scripts/generate_project.py
　dlk/python/dlk/templates/CMakeLists.txt
　dlk/python/dlk/templates/Makefile.tpl
　dlk/python/dlk/templates/include/func/batch_normalization.h
　dlk/python/dlk/templates/include/func/multadd.h
　dlk/python/dlk/templates/src/func/batch_normalization.cpp
　dlk/python/dlk/templates/src/func/multadd.cpp
　dlk/python/dlk/templates/src/network.tpl.cpp

## How has this been tested?
run on X86 / ARM / FPGA

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: dlk/python/dlk/core/optimizer.py [hunk 1]
```diff
@@ -207,6 +227,47 @@ def pass_propagate_quantization_details_into_conv(graph: Graph) -> None:
...
+        )
+
+        graph.add_op(bn_shift_constant)
+        graph.add_op(bn_scale_constant)
+
+        bn_shift_constant.add_output('output', m)
+        bn_scale_constant.add_output('output', m)
+
+        m.add_input('prep_scale', bn_scale_constant)
+        m.add_input('shift', bn_shift_constant)
```
Comment (ruimashita) [inline dlk/python/dlk/core/optimizer.py]: I think it is better to add `Add` and `Mul` operator instead of `BatchNorm` 🤔🤔🤔 
Why do you choice to use `BatchNorm` operator adding `scale` and `shift` ?

Comment (lm-konda) [reply dlk/python/dlk/core/optimizer.py]: At least, current implementation is better than `Mul` and `Add` in terms of memory access.
In addition, it is better to have fuse with Convolution when possible.

**Comment by lm-konda** [comment]
I got a help from @nlpng -san, I add new operator MultAdd and refactor related parts.
Please review again.

**Comment by ruimashita** [comment]
Need to fix same problem as #172 .

**Comment by iizukak** [comment]
This PR has no progress recently, So I close this PR temporary.

**Comment by ruimashita** [comment]
Reopen this PR, I will take over this

File: dlk/python/dlk/templates/src/func/multadd.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#if defined(USE_NEON)
+  const T_UINT size = out_height * out_width;  
+#pragma omp parallel for
+  for (T_UINT f = 0; f < size; f++) {
+
+    T_FLOAT *in_temp = &input[f * out_depth];
+    T_FLOAT *out_temp = &output[f * out_depth];
+
+    T_UINT d = 0;
+    for (; d < out_depth; d += 4) {
```
Comment (n-nez) [inline dlk/python/dlk/templates/src/func/multadd.cpp]: Beware of https://github.com/blue-oil/blueoil/pull/172 :face_with_head_bandage:

**Comment by primenumber** [comment]
This PR is inactive and what to do in this PR was implemented in #658 .
